src/main/scala/cognite/spark/v1/NumericDataPointsRelation.scala

```diff
@@ -147,18 +149,23 @@ class NumericDataPointsRelationV1(config: RelationConfig)(sqlContext: SQLContext
     Row.fromSeq(indicesOfRequiredFields.map(idx => rowOfAllFields.get(idx)))
   }
 
-  def insertSeqOfRows(rows: Seq[Row]): IO[Unit] = {
-    val dataPoints =
-      rows.map(r => fromRow[InsertDataPointsItem](r))
-
-    val updates = dataPoints.groupBy(_.getCogniteId).map {
-      case (id, dataPoints) =>
-        client.dataPoints
-          .insert(id, dataPoints.map(dp => SdkDataPoint(dp.timestamp, dp.value)))
-          .flatTap(_ => incMetrics(itemsCreated, dataPoints.size))
-    }
+  def insertRowIterator(rows: Iterator[Row]): IO[Unit] = {
+    // we basically use Stream.fromIterator instead of Seq.grouped, because it's significantly more efficient
+    val dataPoints = Stream.fromIterator(
+      rows.map(r => fromRow[InsertDataPointsItem](r)),
+      chunkSize = Constants.CreateDataPointsLimit
+    )
 
-    updates.toVector.parSequence_
+    dataPoints.chunks
+      .flatMap(c => Stream.emits(c.toVector.groupBy(_.getCogniteId).toVector))
+      .parEvalMapUnordered(config.parallelismPerPartition) {
+        case (id, dataPoints) =>
+          client.dataPoints
+            .insert(id, dataPoints.map(dp => SdkDataPoint(dp.timestamp, dp.value)))
+            .flatTap(_ => incMetrics(itemsCreated, dataPoints.size))
+      }
+      .compile
+      .drain
```

Comment:
Identical in here and StringDataPointsRelation, I think? Can we deduplicate?

Reply:
Not sure how, it's working with different types (SdkDataPoint, InsertDataPointsItem). I mean it's possible with type arguments and lambda function, but it would not be more readable, IMHO.